server/src/main/java/org/elasticsearch/action/admin/indices/rollover/AutoRolloverAction.java

```diff
@@ -0,0 +1,116 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.action.admin.indices.rollover;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.MetadataDataStreamsService;
+import org.elasticsearch.cluster.routing.allocation.AllocationService;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.tasks.CancellableTask;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Map;
+
+/**
+ * Api that automatically rolls over a data stream that has the flag {@link DataStream#rolloverOnWrite()} enabled. These requests always
+ * originate from requests that write into the data stream.
+ */
+public final class AutoRolloverAction extends ActionType<RolloverResponse> {
+
+    public static final NodeFeature DATA_STREAM_AUTO_ROLLOVER = new NodeFeature("data_stream.rollover.auto");
+
+    public static final AutoRolloverAction INSTANCE = new AutoRolloverAction();
+    public static final String NAME = "indices:admin/data_stream/auto_rollover";
+
+    private AutoRolloverAction() {
+        super(NAME);
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    public static final class TransportAction extends TransportRolloverAction {
+
+        @Inject
+        public TransportAction(
+            TransportService transportService,
+            ClusterService clusterService,
+            ThreadPool threadPool,
+            ActionFilters actionFilters,
+            IndexNameExpressionResolver indexNameExpressionResolver,
+            MetadataRolloverService rolloverService,
+            AllocationService allocationService,
+            MetadataDataStreamsService metadataDataStreamsService,
+            Client client
+        ) {
+            super(
+                AutoRolloverAction.INSTANCE,
+                transportService,
+                clusterService,
+                threadPool,
+                actionFilters,
+                indexNameExpressionResolver,
+                rolloverService,
+                client,
+                allocationService,
+                metadataDataStreamsService
+            );
+        }
+
+        @Override
+        protected void masterOperation(
+            Task task,
+            RolloverRequest rolloverRequest,
```

Comment:
Should we add a check (not just an assert, but maybe an actual exception check) that the `rolloverRequest` doesn't have any conditions here? I think it might be a good sanity check.

Reply:
Since this is an internal operation I am thinking of using an `IllegalStateException`.

Reply:
> Update: maybe even better create a new Request for this that only accepts a target name.

This won't work because we are extending the `TransportRolloverAction`.

Reply:
So going back to:
- asserting that the request does not set parameters that shouldn't be set (to inform the dev in time)
- create a new request when submitting the rollover task and that ensures that the rollover will happen.